### ISSUE TITLE
Inaccuracy of jax.numpy.sum

### ISSUE THREAD

**Issue by nickmcgreivy** [google/jax#4199]
So when I run this code:
```python
import numpy as np
import jax.numpy as jnp
from jax.config import config
config.update("jax_enable_x64", True)

N = int(1e7)
matrix = np.asarray([np.zeros(N), np.zeros(N), np.ones(N) / N])
jmatrix = jnp.asarray([jnp.zeros(N), jnp.zeros(N), jnp.ones(N) / N])

print(np.sum(matrix) - 1.0)
print(jnp.sum(jmatrix) - 1.0)
print(np.sum(matrix[2, :]) - 1.0)
print(jnp.sum(jmatrix[2, :]) - 1.0)
print(np.sum(matrix, axis=1)[2] - 1.0)
print(jnp.sum(jmatrix, axis=1)[2] - 1.0)
```
I get
```
1.7319479184152442e-14
1.2276624161700056e-11
1.7319479184152442e-14
-9.992007221626409e-16
1.7319479184152442e-14
-9.992007221626409e-16
```
It seems this has to do with [partial pairwise summation](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.sum.html). I see #3517, but that appears to be a slightly different issue. I have two questions:

1. The docs say 

> This improved precision is always provided when no axis is given. 

But for `jnp`, if I give no axis, the precision is terrible (1e-11). If I give the axis, or I map over a vector, Jax does better. Do you agree the docs seem to contradict this example?

2. Is there some way I can predict when these errors might arise in advance? I'd love to be able to use `jnp.sum` without worrying about the whether the result will be sufficiently numerically precise.

**Comment by jakevdp** [comment]
Hi – note that the details of the docstring at the link you provided have to do with numpy, not JAX (note at the top where it says "Original docstring below"), and so those details don't actually apply to JAX's implementation .

The precise order of floating point operations performed by JAX will generally depend on the backend implementation, whether the code is JIT-compiled or not, whether you're running code on a CPU, GPU, TPU, etc.

Others may be able to chime in with more detailed information about what precision you should expect in each situation.

**Comment by nickmcgreivy** [comment]
Ah! So this isn't an error in the docs. Thanks for clarifying.

I'm running this on my CPU for reference. Running it on a GPU, I get 

1.7319479184152442e-14
1.7319479184152442e-14
1.7319479184152442e-14
1.7319479184152442e-14
1.7319479184152442e-14
2.0872192862952943e-14```

**Comment by jakevdp** [comment]
Closing, as I don't think there's any action item here.
